### PR TITLE
Deprecate `--nprocs` option for `dask-worker` CLI

### LINKED ISSUE #2471: Remove --nprocs keyword from dask-worker command?
This ends up causing confusion and isn't best practice in production.  I wonder if we should start to deprecate this keyword argument instead.

cc @lesteve @guillaumeeb @jcrist @jacobtomlinson

### PR DESCRIPTION
**Pull request by bryanwweber**
Per #2471, `nprocs` is potentially confusing and should be replaced by `num-workers`.

- [X] Closes #2471
- [x] Tests added / passed
- [X] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by jrbourbeau** [comment]
add to allowlist

**Comment by jrbourbeau** [comment]
cc @guillaumeeb for visibility (as this change will impact `dask-jobqueue`)

File: distributed/cli/dask_worker.py [hunk 1]
```diff
@@ -295,23 +305,27 @@ def main(
         if v is not None
     }
 
-    if nprocs == "auto":
-        nprocs, nthreads = nprocesses_nthreads()
+    if nprocs is not None:
+        warnings.warn("The --nprocs flag has been renamed to --num-workers. ")
```
Comment (jrbourbeau) [inline distributed/cli/dask_worker.py:309]: Since this is a deprecation warning, we should probably raise a `FutureWarning` instead of a `UserWarning`

Comment (jrbourbeau) [reply distributed/cli/dask_worker.py:309]: Let's also mention that `--nprocs` will be removed in a future release in the warning message

File: distributed/cli/tests/test_dask_worker.py [hunk 2]
```diff
@@ -280,6 +282,17 @@ def test_nprocs_expands_name(loop):
                     assert len(set(names)) == 4
 
 
+def test_nprocs_deprecation():
+
+    runner = CliRunner()
+    with pytest.warns(UserWarning, match="nprocs"):
```
Comment (jrbourbeau) [inline distributed/cli/tests/test_dask_worker.py:288]: This checks that the warning message emitted is slightly more meaningful 

```suggestion
    with pytest.warns(FutureWarning, match="renamed to --num-workers"):
```

File: distributed/cli/tests/test_dask_worker.py [hunk 3]
```diff
@@ -280,6 +282,17 @@ def test_nprocs_expands_name(loop):
...
+        try:
+            runner.invoke(distributed.cli.dask_worker.main, ["--nprocs=2"])
+        except ValueError:
+            # didn't pass scheduler
+            pass
```
Comment (jrbourbeau) [inline distributed/cli/tests/test_dask_worker.py:293]: As is, this test checks that a warning is emitted when `--nprocs` is specified, but doesn't check that the CLI option still works. We could instead make this test more similar to `test_num_workers_negative` where we do spin up a scheduler and client and check that when, for example, `--nprocs 2` is passed to `dask-worker` that we not only get the expected deprecation warning, but two workers are also created

File: distributed/cli/dask_worker.py [hunk 4]
```diff
@@ -295,23 +305,27 @@ def main(
         if v is not None
     }
 
-    if nprocs == "auto":
-        nprocs, nthreads = nprocesses_nthreads()
+    if nprocs is not None:
```
Comment (jrbourbeau) [inline distributed/cli/dask_worker.py:308]: Users could specify both `--nprocs` and `--num-workers` at the same time (though this is definitely an edge case). Currently we'll emit a warning and use the `--nprocs` value, which may be confusing. Instead we should raise an error in the case that both `--nprocs` and `--num-workers` is provided, since it's not clear which one should be used.

**Review by jrbourbeau — COMMENTED** [review]
Thanks @bryanwweber! I know this is still a draft PR, but I still wanted to leave some early feedback. 

Also, unfortunately `distributed` has several known flaky tests at the moment. The test failures popping up here are all unrelated to the changes in this PR.

**Comment by bryanwweber** [comment]
@jrbourbeau Sure, thanks for the review! I'll take a look tomorrow. I also noticed that edge case, but wanted to follow the pre-existing convention for the `bokeh`/`dashboard` options. I'll be happy to update.

I just pushed 3 more commits to catch the few other places where `nprocs` was used. I'm not sure if I should also update the `old_ssh` interface, so I put those changes into a separate commit to make it easy to drop.

**Comment by guillaumeeb** [comment]
> cc @guillaumeeb for visibility (as this change will impact `dask-jobqueue`)

Thanks @jrbourbeau, I'll try to follow the evolution here!

**Comment by bryanwweber** [comment]
@jrbourbeau I can't reproduce this test failure locally: https://github.com/dask/distributed/runs/4791743011?check_suite_focus=true#step:12:182 possibly because I'm on a mac and this is a Windows failure?

**Review by jacobtomlinson — APPROVED** [review]
The SSH changes look reasonable to me.

This will be breaking for `old_ssh` but perhaps we should think about removing that now anyway.

**Comment by jacobtomlinson** [comment]
Sorry for the confusion with the review dismissal. I'm happy with this change. 

We should remove `old_ssh` at some point, but that requires migrating the CLI over to use it.

File: distributed/deploy/ssh.py [hunk 5]
```diff
@@ -89,7 +89,7 @@ def __init__(
         self.kwargs = copy.copy(kwargs)
         self.name = name
         self.remote_python = remote_python
-        self.nprocs = self.kwargs.pop("nprocs", 1)
+        self.num_workers = self.kwargs.pop("num_workers", 1)
```
Comment (bryanwweber) [inline distributed/deploy/ssh.py:92]: @jrbourbeau I just realized that this is a breaking change if anyone is using the `Worker` class directly. I'll add similar deprecation checks here as elsewhere.

File: distributed/cli/tests/test_dask_ssh.py [hunk 6]
```diff
@@ -1,9 +1,24 @@
...
     runner = CliRunner()
     result = runner.invoke(main, ["--version"])
     assert result.exit_code == 0
+
+
+def test_nprocs_renamed_to_num_workers():
+    with popen(["dask-ssh", "--nprocs=2", "127.0.0.1"]) as cluster:
+        assert any(
+            b"renamed to --num-workers" in cluster.stderr.readline() for i in range(15)
+        )
```
Comment (bryanwweber) [inline distributed/cli/tests/test_dask_ssh.py:17]: This test is timing out on Python 3.7 and 3.8, but curiously not on 3.9. I've set up a Docker container to try and debug it locally.

File: distributed/deploy/tests/test_old_ssh.py [hunk 7]
```diff
@@ -28,3 +28,25 @@ def test_cluster(loop):
...
+
+def test_num_workers_with_nprocs_is_an_error():
+    with pytest.raises(ValueError, match="Both nprocs and num_workers"):
+        SSHCluster(
+            scheduler_addr="127.0.0.1",
+            scheduler_port=8687,
+            worker_addrs=(),
+            nprocs=2,
+            num_workers=2,
+        )
```
Comment (bryanwweber) [inline distributed/deploy/tests/test_old_ssh.py:52]: This test is failing, perhaps because I removed the `async` bits here? I'll try in my Docker container to reproduce the failures and fix it.

**Comment by bryanwweber** [comment]
@jrbourbeau I tried to use `Client` as we discussed to test `dask_ssh` but the test always timesout. I wonder if this is the cause: https://github.com/dask/distributed/blob/f4451b66cbb5c365d92a79394440877e70095688/distributed/cli/dask_ssh.py#L216-L217

I suspect this might be why I needed to issue the SIGINT to get any output, if that function is blocking until SIGINT is sent?

**Comment by graingert** [comment]
> @jrbourbeau I tried to use `Client` as we discussed to test `dask_ssh` but the test always times out. I wonder if this is the cause:
> 
> https://github.com/dask/distributed/blob/f4451b66cbb5c365d92a79394440877e70095688/distributed/cli/dask_ssh.py#L216-L217
> 
> I suspect this might be why I needed to issue the SIGINT to get any output, if that function is blocking until SIGINT is sent?

I suspect this is because the stdout stream is block-buffered by default (when non-interactive) you could set PYTHONUNBUFFERED if you want the output as soon as possible.

https://docs.python.org/3/library/sys.html#sys.stderr

**Comment by bryanwweber** [comment]
@graingert Thanks for the thought! I don't think that's the problem here, for a few reasons. The modified test code with `Client` looks like:

```python
def test_ssh_cli_nprocs_renamed_to_num_workers(loop):
    num_workers = 2
    with popen(
        ["dask-ssh", f"--nprocs={num_workers}", "--nohost", "localhost"]
    ) as cluster:
        with Client("tcp://127.0.0.1:8786", loop=loop) as c:
            c.wait_for_workers(num_workers, timeout="15 seconds")
```

1. The timeout failure happens due to `pytest-timeout`, not because the `Client` has timed out. I'm running pytest as `pytest -s distributed/cli/tests/test_dask_ssh.py::test_ssh_cli_nprocs_renamed_to_num_workers --timeout=30`, and the timeout only happens after 30 seconds, not 15.
2. The `-s` flag to `pytest` should result in unbuffered output, if I understand correctly
3. I also tried modifying the `bufsize` kwarg to `popen` to set it to 0, which also did not result in any output
4. The modified test is not actually testing or waiting on any output to `stdout` or `stderr`, but it still times out

**Comment by jsignell** [comment]
Sorry to be super late to the conversation, but is there a good reason to use `num_` rather than `n` or `n_` as the prefix? And if there is, can we consolidate `nthreads` and `n_workers` to this prefix?

**Comment by bryanwweber** [comment]
> Sorry to be super late to the conversation, but is there a good reason to use `num_` rather than `n` or `n_` as the prefix? And if there is, can we consolidate `nthreads` and `n_workers` to this prefix?

No problem @jsignell! `num_workers` was suggested in the linked issue, https://github.com/dask/distributed/issues/2471#issuecomment-456216327. `n_workers` would be fine with me 😄 

I believe `nthreads` and `num_workers` are orthogonal. `nthreads` sets the number of threads per process, and `num_workers` sets the number of processes. But, as I'm new to the project, I could be mistaken 😃

**Comment by bryanwweber** [comment]
Thanks @jsignell! Do you prefer `nworkers` and `nthreads` or with the underscore? The advantage of `nworkers` is that it won't require a deprecation cycle with `nthreads`, but my personal preference is for the underscores.

**Comment by jsignell** [comment]
> Thanks @jsignell! Do you prefer `nworkers` and `nthreads` or with the underscore? The advantage of `nworkers` is that it won't require a deprecation cycle with `nthreads`, but my personal preference is for the underscores.

No preference. There is an `n_workers` already elsewhere (at least in LocalCluster, but possibly other places). Not sure if we should match or be different for greppability (probably match).

**Comment by bryanwweber** [comment]
Whoops, sorry @jacobtomlinson I hit the wrong button there. Feel free to dismiss the review request (or do a review if you want to 😄 )

**Comment by bryanwweber** [comment]
@jsignell Based on a conversation with @jrbourbeau we agreed that using `--nworkers` for CLI flags and `n_workers` in Python code would be the most self-consistent approach.

**Comment by bryanwweber** [comment]
@jrbourbeau I _think_ I tracked down the SSH failures to the `Client` not having enough time on the CI machines to connect to the scheduler. I bumped that timeout, hopefully it will fix it 🤞

**Comment by bryanwweber** [comment]
@jrbourbeau The test failing on Ubuntu/3.9/no ci1 doesn't look like it's due to my change, but I can't quite tell. Do you have any insight? https://github.com/dask/distributed/runs/4968972824?check_suite_focus=true#step:12:1434 Otherwise, waiting on macOS like some other PRs

File: distributed/deploy/old_ssh.py [hunk 8]
```diff
@@ -349,12 +350,27 @@ def __init__(
         nanny_port=None,
         remote_dask_worker="distributed.cli.dask_worker",
         local_directory=None,
+        **kwargs,
```
Comment (jrbourbeau) [inline distributed/deploy/old_ssh.py:353]: Adding `**kwargs` here will cause invalid keyword arguments to go unnoticed. For example, today a typo of `memory_limt` instead of `memory_limit` will raise an error, while with this change as is such a typo will be silently ignored.

Instead I'd rather us either keep an explicit `nprocs=` keyword around, but still raise a `ValueError` / `FutureWarning` accordingly, or do something like (below is pseudocode)

```python
nprocs = kwargs.pop("nprocs")
if kwargs:
    raise TypeError("...")
<existing ValueError / FutureWarning logic>

File: distributed/deploy/old_ssh.py [hunk 9]
```diff
@@ -349,12 +350,27 @@ def __init__(
...
+            warnings.warn(
+                "The nprocs argument will be removed in a future release. It has been "
+                "renamed to n_workers.",
+                FutureWarning,
+            )
+            n_workers = kwargs["nprocs"]
+        elif n_workers is None:
+            n_workers = 1
+
+        self.n_workers = n_workers
```
Comment (jrbourbeau) [inline distributed/deploy/old_ssh.py:373]: Can we add a small `self.nprocs` property which redirects users to `self.n_workers` instead?

File: distributed/deploy/ssh.py [hunk 10]
```diff
@@ -89,7 +89,19 @@ def __init__(
...
+            )
+        elif kwargs.get("nprocs") is not None:
+            warnings.warn(
+                "The nprocs argument will be removed in a future release. It has been "
+                "renamed to n_workers.",
+                FutureWarning,
+            )
+            self.n_workers = self.kwargs.pop("nprocs")
+        else:
+            self.n_workers = self.kwargs.pop("n_workers", 1)
```
Comment (jrbourbeau) [inline distributed/deploy/ssh.py:104]: Similar comment here about adding a `self.nprocs` property redirecting to `self.n_workers`

File: distributed/deploy/ssh.py [hunk 11]
```diff
@@ -89,7 +89,19 @@ def __init__(
...
+            raise ValueError(
+                "Both nprocs and n_workers were specified. Use n_workers only."
+            )
+        elif kwargs.get("nprocs") is not None:
+            warnings.warn(
+                "The nprocs argument will be removed in a future release. It has been "
+                "renamed to n_workers.",
+                FutureWarning,
+            )
+            self.n_workers = self.kwargs.pop("nprocs")
```
Comment (jrbourbeau) [inline distributed/deploy/ssh.py:102]: Not sure if this is needed or not

```suggestion
            self.n_workers = self.kwargs.pop("nprocs", 1)
```

**Review by jrbourbeau — APPROVED** [review]
Thanks for all your efforts on this @bryanwweber. Agreed those test failures are unrelated to the changes here. I left a few final review comments, but this looks good overall

File: distributed/deploy/old_ssh.py (hunk 9 above)
Comment (bryanwweber) [reply distributed/deploy/old_ssh.py:373]: I'm happy to add this. However, setting `nprocs` or `n_workers` as an attribute (after `__init__()` runs) doesn't actually change anything about the cluster, unless the user subsequently runs `add_worker()`. For that matter, it's not all that clear to me what the behavior should be... do all the worker nodes need to be updated with the number of worker processes? Is there a method to do that (there doesn't seem to be from my quick read of the code)?

File: distributed/deploy/ssh.py (hunk 10 above)
Comment (bryanwweber) [reply distributed/deploy/ssh.py:104]: Similar response as above 😄

File: distributed/deploy/ssh.py (hunk 11 above)
Comment (bryanwweber) [reply distributed/deploy/ssh.py:102]: It shouldn't be needed, but wouldn't hurt to add it.

**Review by jrbourbeau — APPROVED** [review]
Thanks @bryanwweber! This is in
